compiler/src/iree/compiler/Codegen/Dialect/Codegen/IR/IREECodegenDialect.cpp

```diff
@@ -79,73 +81,63 @@ static LogicalResult validateTuningEntrypoint(
     return kernelConfigOp.emitError()
            << "Missing attribute '" << kTuningSpecEntrypointAttrName
            << "' in named sequence '" << kKernelConfigSpecName << "'"
-           << kRequiredByDefaultAttr;
+           << requiredByDefaultAttrMessage;
   }
 
   if (numTuningEntryPoints != 1) {
     return moduleOp.emitError()
            << "Expected one named sequence with '"
            << kTuningSpecEntrypointAttrName << "', but found "
-           << numTuningEntryPoints << kRequiredByDefaultAttr;
+           << numTuningEntryPoints << requiredByDefaultAttrMessage;
   }
 
   return success();
 }
 
 LogicalResult
 validateKernelConfigContents(transform::NamedSequenceOp kernelConfigOp,
-                             const std::string &kRequiredByDefaultAttr) {
-  transform::ForeachMatchOp foreachMatchOp;
-  bool hasYieldOp = false;
-  for (Block &block : kernelConfigOp.getBlocks()) {
-    for (Operation &op : block) {
-      if (auto foreachOp = dyn_cast<transform::ForeachMatchOp>(op)) {
-        if (foreachMatchOp) {
-          return kernelConfigOp.emitError()
-                 << "'" << kKernelConfigSpecName
-                 << "' must contain exactly one 'ForeachMatchOp'"
-                 << kRequiredByDefaultAttr;
-        }
-        foreachMatchOp = foreachOp;
-      } else if (isa<transform::YieldOp>(op)) {
-        if (hasYieldOp) {
-          return kernelConfigOp.emitError()
-                 << "'" << kKernelConfigSpecName
-                 << "' must contain exactly one 'transform::YieldOp'"
-                 << kRequiredByDefaultAttr;
-        }
-        hasYieldOp = true;
-      } else {
-        return kernelConfigOp.emitError()
-               << "Unexpected op '" << op.getName() << "' in '"
-               << kKernelConfigSpecName << "'" << kRequiredByDefaultAttr;
-      }
-    }
+                             StringRef requiredByDefaultAttrMessage) {
+  // Ensure there is exactly one block.
+  if (kernelConfigOp.getBlocks().size() != 1) {
+    return kernelConfigOp.emitError()
+           << "'" << kKernelConfigSpecName << "' must contain exactly one block"
+           << requiredByDefaultAttrMessage;
   }
 
-  // Ensure both required operations are present.
+  Block &block = kernelConfigOp.getBlocks().front();
+  // Ensure there are exactly two operations.
+  if (block.getOperations().size() != 2) {
```

Comment:
Use `hasNItems` from https://llvm.org/doxygen/STLExtras_8h.html for size checks in linked lists. This way it won't iterate over the whole thing once you know there are more than 2 instructions.